### DISCUSSION TITLE
KAD BucketInserts::OnConnected works as not expected

### DISCUSSION THREAD

**Discussion by gurinderu** [libp2p/rust-libp2p#4888] [Q&A]
I have made a simple test and I expect to see the insertion in the k-bucket after connection established event. I have checked the code and see the calling of connection_updated only for the on_connection_closed function. Is it bug or expected behaviour ?

```#[test]
fn simple_check() {
    let filter = tracing_subscriber::EnvFilter::builder()
        .with_default_directive(Level::INFO.into()).from_env_lossy();

    tracing_subscriber::fmt()
        .with_env_filter(filter)
        .try_init()
        .ok();

    let (_swarm_1_addr, mut swarm_1) = build_node_with_config(Config::default());
    let (swarm_2_addr, swarm_2) = build_node_with_config(Config::default());


    swarm_1.dial(swarm_2_addr).expect("Could not dial");

    let mut swarms = vec![swarm_1, swarm_2];

    block_on(poll_fn(move |ctx| {
        for (_i, swarm) in swarms.iter_mut().enumerate() {
            loop {
                match swarm.poll_next_unpin(ctx) {
                    Poll::Ready(Some(SwarmEvent::Behaviour(Event::RoutingUpdated {
                                                               peer,
                                                               addresses, ..
                                                           }))) => {
                        tracing::info!("RoutingUpdated {:?} {:?}", peer, addresses);
                    }
                    Poll::Ready(_) => {}
                    Poll::Pending => break,
                }
            }
        }
        Poll::<()>::Pending
    }));
}```

**Comment by thomaseizinger** [comment]
Is `swarm2` operating in server-mode? If it is not operating in server-mode, it will not advertise the kademlia protocol and thus swarm1 will not add it to the routing table.

Server-mode is automatically activated once a node learns its external address. These can either be configured using `Swarm::add_external_address` or another behaviour issues a `ToSwarm::ExternalAddrConfirmed` event.

As a last resort, you can also force the mode using `kad::Behaviour::set_mode`. I'd discourage that though because for your node to be a useful kademlia node, it _must_ have an external address, otherwise other nodes will not be able to connect to it, so why should it be in the routing table in the first place? It is better if you correctly configure the external address of your node or use a behaviour that determines that.

For example, `libp2p-identify` will report you the _observed_ address of your node using `SwarmEvent::NewExternalAddrCandidate`. Depending on your network configuration, an external addr candidate may or may not be an external address, e.g. if your node is behind NAT and you don't have port-forwarding enabled, the observed address won't be dialable. But in that case, the node should also not be in kademlia server-mode because it isn't actually reachable for other nodes.

We are currently working on an improved version of AutoNAT ([v2](https://github.com/libp2p/rust-libp2p/issues/4524)) that will make it possible to automatically detect your NAT status.

You can use the v1 version in the meantime but it has some bugs unfortunately.

**Reply by gurinderu** [reply]
Both swarms are in the server mode

**Reply by gurinderu** [reply]
log output (sure custom log, because the kad behaviour doesn't have logging:( )
```
2023-11-20T09:58:02.504483Z  INFO libp2p_swarm: local_peer_id=12D3KooWGSBZfGBuAX61sJkQu5MbvJYk2zZXURst428YcXWdK6CU
2023-11-20T09:58:02.506077Z  INFO libp2p_swarm: local_peer_id=12D3KooWQsP5RurSu4NHV1mBVu1MArdBKst6t7QW13T3hWXmEpGW
2023-11-20T09:58:02.512082Z  INFO libp2p_kad::behaviour: on_connection_established 12D3KooWQsP5RurSu4NHV1mBVu1MArdBKst6t7QW13T3hWXmEpGW [] 0
2023-11-20T09:58:02.512173Z  INFO libp2p_kad::behaviour: on_connection_established 12D3KooWGSBZfGBuAX61sJkQu5MbvJYk2zZXURst428YcXWdK6CU [] 0
2023-11-20T09:58:12.515259Z  INFO libp2p_kad::behaviour: on_connection_closed PeerId("12D3KooWQsP5RurSu4NHV1mBVu1MArdBKst6t7QW13T3hWXmEpGW") 0 ConnectionId(1)
2023-11-20T09:58:12.515456Z  INFO libp2p_kad::behaviour: connection_updated PeerId("12D3KooWQsP5RurSu4NHV1mBVu1MArdBKst6t7QW13T3hWXmEpGW") None Disconnected
2023-11-20T09:58:12.515780Z  INFO libp2p_kad::behaviour: on_connection_closed PeerId("12D3KooWGSBZfGBuAX61sJkQu5MbvJYk2zZXURst428YcXWdK6CU") 0 ConnectionId(2)
2023-11-20T09:58:12.515871Z  INFO libp2p_kad::behaviour: connection_updated PeerId("12D3KooWGSBZfGBuAX61sJkQu5MbvJYk2zZXURst428YcXWdK6CU") None Disconnected
```
As you can see we don't have buckets changing events, only on closing connection. But I expect to see them on the connection established. At least for Bucket Inserts::OnConnected which is a default.

**Reply by thomaseizinger** [reply]
> because the kad behaviour doesn't have logging

It does but we use `tracing` so you need to create a `tracing_subscriber`.

**Reply by thomaseizinger** [reply]
Do you use the identify behaviour? Are you making any kademlia requests (like bootstrap?) on your network?

We can't magically know whether the other node speaks kademlia and thus add it to the routing table.

We only learn about the other node being a kademlia node:

a) via identify
b) through a successful kademlia stream opening (either inbound or outbound)

**Reply by gurinderu** [reply]
I already use tracing_subscriber :)

**Reply by gurinderu** [reply]
Let me check the behaviour with the identify
